modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningPredicate.java

```diff
@@ -0,0 +1,197 @@
+                NodeWithConsistencyToken nodeWithConsistencyToken = colocationGroup.assignments().get(p);
+                if (!remainingPartitions.contains(p)) {
+                    continue;
+                }
+
+                if (Objects.equals(nodeName, nodeWithConsistencyToken.name())) {
+                    long t = nodeWithConsistencyToken.enlistmentConsistencyToken();
+
+                    partsWithConsistencyTokens.add(new PartitionWithConsistencyToken(p, t));
+                    newNodes.add(nodeName);
+                }
+            }
+
+            if (!partsWithConsistencyTokens.isEmpty()) {
+                partitionsPerNode.put(nodeName, partsWithConsistencyTokens);
+            }
+        }
+
+        // Keep assignments intact, because they are used by DestinationFactory.
+        return new ColocationGroup(
+                colocationGroup.sourceIds(),
+                List.copyOf(newNodes),
+                colocationGroup.assignments(),
+                partitionsPerNode
+        );
+    }
+
+    @Nullable
+    private static IntSet computeRemainingPartitions(
+            IgniteTable table,
+            PartitionPruningColumns pruningColumns,
+            Object[] dynamicParameters
+    ) {
+        ImmutableIntList keys = table.distribution().getKeys();
+        IntSet remainingPartitions = new IntArraySet(pruningColumns.columns().size());
+        PartitionCalculator partitionCalculator = table.partitionCalculator().get();
+
+        for (Int2ObjectMap<RexNode> columns : pruningColumns.columns()) {
+            for (int key : keys) {
+                RexNode node = columns.get(key);
+                NativeType physicalType = table.descriptor().columnDescriptor(key).physicalType();
+
+                // TODO: Custom data types in partition pruning.
+                if (physicalType.spec() == NativeTypeSpec.UUID) {
+                    return null;
+                }
+
+                Object val = getNodeValue(physicalType, node, dynamicParameters);
+
+                partitionCalculator.append(val);
+            }
+
+            int p = partitionCalculator.partition();
+            remainingPartitions.add(p);
+        }
+
+        return remainingPartitions;
+    }
+
+    @Nullable
+    private static Object getNodeValue(NativeType physicalType, RexNode node, Object[] dynamicParameters) {
+        Object val;
+
+        if (node instanceof RexLiteral) {
+            val = getValueFromLiteral(physicalType, (RexLiteral) node);
+        } else if (node instanceof RexDynamicParam) {
+            RexDynamicParam dynamicParam = (RexDynamicParam) node;
+            val = dynamicParameters[dynamicParam.getIndex()];
+        } else {
+            throw new IllegalArgumentException("Unexpected column value node: " + node);
+        }
+
+        return val;
+    }
+
+    @Nullable
+    private static Object getValueFromLiteral(NativeType physicalType, RexLiteral lit) {
+        if (physicalType.spec() == NativeTypeSpec.DATE) {
```

Comment:
This code can be rewritten by using an expression compilation, but `ExpressionFactory` currently depends on `ExecutionContext` for no reason at all (it uses `TypeFactory` and `RowHandler`, if there are other dependencies I almost certain that they are also trivial). 
It would be better to make `ExpressionFactory` independent of `ExecutionContext`.